File: tests/test_vectors.py
"""Проверка всех тестовых векторов раздела 8 спецификации.

Проверяются и промежуточные величины -- таблицы лабиринта и подключи:
они локализуют ошибку до конкретного этапа, а пораундовая трассировка TV3 --
до конкретного раунда.

Запуск:  python tests/test_vectors.py
"""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import dedalyan as D
from dedalyan_harness import Reporter, fmt_bits, make_parser

# ---- ожидаемые значения из спецификации ----------------------------------

LAB_ZERO = ("a9dc013f46e85b27", "a1e4c03d76b9f825")
LAB_SEQ = ("3bed640f928c157a", "e68359f1b4cd207a")
LAB_APPLY = (0x0123456789AB, 0xE6ED640F92CD)

RC_TABLE = [
    0x3188EBE1E0B8, 0xCFC0659B6002, 0x6DF7DF54DF4C, 0x0C2F590E5E96,
    0xAA66D2C7DDE0, 0x489E4C815D2A, 0xE6D5C63ADC74, 0x850D3FF45BBE,
    0x2344B9ADDB08, 0xC17C33675A52, 0x5FB3AD20D99C, 0xFDEB26DA58E6,
    0x9C22A093D830, 0x3A5A1A4D577A, 0xD8919406D6C4, 0x76C90DC0560E,
]

KS_ZERO = [
    0x9864A848B6D4, 0x66845D0326D1, 0x9CE6EFF8CD60, 0xDE534C0AEEC3,
    0x84ED264AC5BB, 0xAEDD6A826C93, 0x2C4D8A772D04, 0xABA7EEC66F0F,
    0xFF822A71A117, 0x13776BBC1FBF, 0x499DA61DEBDF, 0xFF1503BD8AAC,
    0xDC17510F5FAE, 0x3A08B3DAB4F4, 0x63BBB6A9252B, 0x30A6E63F94E6,
]

KS_SEQ = [
    0x352B543E645D, 0x3BB3B319FA45, 0xBA34EE3B5816, 0xF5015D474FAA,
    0x18454389852B, 0x545DAA0A6112, 0x51F0A63D8E60, 0xB9BDDDA2E38C,
    0x4A92ED2D2816, 0x47CDD4E8E543, 0xE10755F1FC74, 0xB7F0A3BD6E0C,
    0x3D5D558B3B7F, 0xE04B6A7BD904, 0x558C5A714B1B, 0xF9958A46E7BF,
]

KEY_SEQ = int.from_bytes(bytes(range(32)), "big")
KEY_ONES = (1 << 256) - 1

# (имя, P, K, C)
TEST_VECTORS = [
    ("TV1", 0x000000000000000000000000, 0, 0x70A4CEAA4A6737FB294A0EDF),
    ("TV2", 0x000000000000000000000001, 0, 0x85A903C2A6B73B50F00E1405),
    ("TV3", 0x0123456789ABCDEF01234567, KEY_SEQ, 0x9B631FE623F15016CDBA801E),
    ("TV4", (1 << 96) - 1, KEY_ONES, 0xA936A309096319F869F6549D),
]

# Раздел 8.5: (F, L, R) после каждого раунда для TV3.
TV3_TRACE = [
    (0x229E7F0EB588, 0xCDEF01234567, 0x23BD3A693C23),
    (0x94E01BC2AAD8, 0x23BD3A693C23, 0x590F1AE1EFBF),
    (0xA9C66A1086E7, 0x590F1AE1EFBF, 0x8A7B5079BAC4),
    (0xFD9E1674955A, 0x8A7B5079BAC4, 0xA4910C957AE5),
    (0xA7C66B61B1B1, 0xA4910C957AE5, 0x2DBD3B180B75),
    (0x82336D2325BE, 0x2DBD3B180B75, 0x26A261B65F5B),
    (0x05734BE79F0D, 0x26A261B65F5B, 0x28CE70FF9478),
    (0x7E3FEBC6F8B6, 0x28CE70FF9478, 0x589D8A70A7ED),
    (0x267030894057, 0x589D8A70A7ED, 0x0EBE4076D42F),
    (0xC572310F4B4C, 0x0EBE4076D42F, 0x9DEFBB7FECA1),
    (0x5F51EC67DD59, 0x9DEFBB7FECA1, 0x51EFAC110976),
    (0xBD7201F955D8, 0x51EFAC110976, 0x209DBA86B979),
    (0x60E553C6F23F, 0x209DBA86B979, 0x310AFFD7FB49),
    (0x1FB2103A188A, 0x310AFFD7FB49, 0x3F2FAABCA1F3),
    (0xAA69E031D8B8, 0x3F2FAABCA1F3, 0x9B631FE623F1),
    (0x6F39670621ED, 0x9B631FE623F1, 0x5016CDBA801E),
]

SHIFTS_SPEC = [45, 47, 1, 3, 5, 7, 9, 11, 13, 15, 17, 19,
               21, 23, 25, 27, 29, 31, 33, 35]

SELECTOR_BITS_SPEC = [26, 30, 34, 38, 42, 46, 2, 6, 10, 14, 18, 22]


def tbl(T) -> str:
    return "".join("%x" % v for v in T)


def main() -> int:
    parser = make_parser(__doc__.splitlines()[0])
    parser.parse_args()
    r = Reporter("Dedalyan test vectors (spec section 8)")

    # ---- раздел 2 --------------------------------------------------------
    r.section("2. Round constants RC_i = ((i + 44) * PHI) mod 2^48")
    bad = [i for i in range(16) if D.round_constant(i) != RC_TABLE[i]]
    r.check(not bad, "RC table matches spec",
            "" if not bad else f"mismatch at i={bad}")

    r.section("7. Rotation amounts r_i and 6.3 selector bits b_j")
    r.check(list(D.SHIFTS) == SHIFTS_SPEC, "r_i for i = -4..15",
            f"got {list(D.SHIFTS)[:4]}...")
    r.check(list(D.SELECTOR_BITS) == SELECTOR_BITS_SPEC, "b_j for j = 0..11")

    # ---- раздел 8.1 ------------------------------------------------------
    r.section("8.1. Labyrinth")
    T0, T1 = D.build_labyrinth(0x0000000000000000)
    r.check(tbl(T0) == LAB_ZERO[0], "T0 for KL = 0", tbl(T0))
    r.check(tbl(T1) == LAB_ZERO[1], "T1 for KL = 0", tbl(T1))

    T0, T1 = D.build_labyrinth(0x0123456789ABCDEF)
    r.check(tbl(T0) == LAB_SEQ[0], "T0 for KL = 0123456789ABCDEF", tbl(T0))
    r.check(tbl(T1) == LAB_SEQ[1], "T1 for KL = 0123456789ABCDEF", tbl(T1))

    got = D.apply_labyrinth(LAB_APPLY[0], (T0, T1))
    r.check(got == LAB_APPLY[1], "Lab(0x0123456789AB)", fmt_bits(got))

    # Обе таблицы обязаны быть перестановками -- иначе лабиринт необратим
    # по построению и Фишер--Йетс реализован неверно.
    r.check(sorted(T0) == list(range(16)) and sorted(T1) == list(range(16)),
            "T0 and T1 are permutations of 0..15")

    # ---- разделы 8.2 и 8.3 -----------------------------------------------
    r.section("8.2/8.3. Key schedule")
    ks = D.key_schedule(0)
    bad = [i for i in range(16) if ks[i] != KS_ZERO[i]]
    r.check(not bad, "subkeys for K = 0",
            "" if not bad else f"mismatch at k[{bad[0]}] = {fmt_bits(ks[bad[0]])}")

    ks = D.key_schedule(KEY_SEQ)
    bad = [i for i in range(16) if ks[i] != KS_SEQ[i]]
    r.check(not bad, "subkeys for K = 000102..1e1f",
            "" if not bad else f"mismatch at k[{bad[0]}] = {fmt_bits(ks[bad[0]])}")

    r.check(all(0 <= k <= D.M for k in ks), "all subkeys fit in 48 bits")

    # ---- раздел 8.4 ------------------------------------------------------
    r.section("8.4. Encryption vectors")
    for name, P, K, C in TEST_VECTORS:
        got = D.encrypt_block(P, K)
        r.check(got == C, f"{name} encrypt", f"{got:024x}")
        back = D.decrypt_block(C, K)
        r.check(back == P, f"{name} decrypt", f"{back:024x}")

    # ---- раздел 8.5 ------------------------------------------------------
    r.section("8.5. Per-round trace for TV3 (localises errors to one round)")
    trace = D.encrypt_block_trace(TEST_VECTORS[2][1], TEST_VECTORS[2][2])
    first_bad = None
    for i, ((f, l, rr), (ef, el, er)) in enumerate(zip(trace, TV3_TRACE)):
        if (f, l, rr) != (ef, el, er):
            first_bad = (i, f, l, rr, ef, el, er)
            break
    if first_bad is None:
        r.check(True, "all 16 rounds match")
    else:
        i, f, l, rr, ef, el, er = first_bad
        r.check(False, f"round {i} mismatch",
                f"got F={fmt_bits(f)} L={fmt_bits(l)} R={fmt_bits(rr)} / "
                f"want F={fmt_bits(ef)} L={fmt_bits(el)} R={fmt_bits(er)}")

    # ---- сверка с C ------------------------------------------------------
    r.section("C backend parity")
    from dedalyan_c import backend
    if not backend.available:
        r.warn("C backend not built", "skipping parity checks; run build.ps1")
    else:
        ok = True
        for name, P, K, C in TEST_VECTORS:
            ctx = backend.new_ctx(D.key_to_bytes(K))
            ok &= backend.encrypt_block(ctx, P) == C
            ok &= backend.decrypt_block(ctx, C) == P
        r.check(ok, "C reproduces all four vectors")

        cT = backend.build_labyrinth(0x0123456789ABCDEF)
        r.check(tbl(cT[0]) == LAB_SEQ[0] and tbl(cT[1]) == LAB_SEQ[1],
                "C labyrinth matches spec")
        r.check(backend.key_schedule(D.key_to_bytes(KEY_SEQ)) == KS_SEQ,
                "C key schedule matches spec")

        ctx = backend.new_ctx(D.key_to_bytes(KEY_SEQ))
        ctrace = backend.trace(ctx, TEST_VECTORS[2][1])
        r.check(ctrace == TV3_TRACE, "C per-round trace matches spec")

    return r.summary()


if __name__ == "__main__":
    sys.exit(main())

File: c/test_vectors.c
/* test_vectors.c -- проверка C-реализации по разделу 8 спецификации.
 * Вывод намеренно на английском: русский ломается в консоли Windows. */

#include "dedalyan.h"
#include "kernels.h"
#include <stdio.h>
#include <string.h>

static int failures = 0;
static int checks = 0;

static void ok(int cond, const char *what)
{
    checks++;
    if (!cond) { failures++; printf("  FAIL  %s\n", what); }
}

static void ok_u64(uint64_t got, uint64_t want, const char *what)
{
    checks++;
    if (got != want) {
        failures++;
        printf("  FAIL  %-34s got %012llx want %012llx\n", what,
               (unsigned long long)got, (unsigned long long)want);
    }
}

static void hex2bytes(const char *hex, uint8_t *out, size_t n)
{
    size_t i;
    for (i = 0; i < n; i++) {
        unsigned v;
        sscanf(hex + 2 * i, "%2x", &v);
        out[i] = (uint8_t)v;
    }
}

/* ---- раздел 2: раундовые константы ------------------------------------- */

static const uint64_t RC_REF[16] = {
    UINT64_C(0x3188EBE1E0B8), UINT64_C(0xCFC0659B6002),
    UINT64_C(0x6DF7DF54DF4C), UINT64_C(0x0C2F590E5E96),
    UINT64_C(0xAA66D2C7DDE0), UINT64_C(0x489E4C815D2A),
    UINT64_C(0xE6D5C63ADC74), UINT64_C(0x850D3FF45BBE),
    UINT64_C(0x2344B9ADDB08), UINT64_C(0xC17C33675A52),
    UINT64_C(0x5FB3AD20D99C), UINT64_C(0xFDEB26DA58E6),
    UINT64_C(0x9C22A093D830), UINT64_C(0x3A5A1A4D577A),
    UINT64_C(0xD8919406D6C4), UINT64_C(0x76C90DC0560E)
};

static void test_round_constants(void)
{
    int i;
    printf("[8.0] round constants\n");
    for (i = 0; i < 16; i++) ok_u64(dedalyan_rc((unsigned)i), RC_REF[i], "RC_i");
}

/* ---- раздел 8.1: лабиринт ---------------------------------------------- */

static void test_labyrinth(void)
{
    static const char *T0_ZERO = "a9dc013f46e85b27";
    static const char *T1_ZERO = "a1e4c03d76b9f825";
    static const char *T0_SEQ  = "3bed640f928c157a";
    static const char *T1_SEQ  = "e68359f1b4cd207a";
    uint8_t T[2][16];
    char buf[17];
    int i;

    printf("[8.1] labyrinth\n");

    dedalyan_build_labyrinth(UINT64_C(0), T);
    for (i = 0; i < 16; i++) buf[i] = "0123456789abcdef"[T[0][i]];
    buf[16] = 0; ok(strcmp(buf, T0_ZERO) == 0, "T0 for KL=0");
    for (i = 0; i < 16; i++) buf[i] = "0123456789abcdef"[T[1][i]];
    buf[16] = 0; ok(strcmp(buf, T1_ZERO) == 0, "T1 for KL=0");

    dedalyan_build_labyrinth(UINT64_C(0x0123456789ABCDEF), T);
    for (i = 0; i < 16; i++) buf[i] = "0123456789abcdef"[T[0][i]];
    buf[16] = 0; ok(strcmp(buf, T0_SEQ) == 0, "T0 for KL=0123456789ABCDEF");
    for (i = 0; i < 16; i++) buf[i] = "0123456789abcdef"[T[1][i]];
    buf[16] = 0; ok(strcmp(buf, T1_SEQ) == 0, "T1 for KL=0123456789ABCDEF");

    ok_u64(dedalyan_apply_labyrinth(UINT64_C(0x0123456789AB),
                                    (const uint8_t (*)[16])T),
           UINT64_C(0xE6ED640F92CD), "Lab(0x0123456789AB)");
}

/* ---- разделы 8.2 и 8.3: подключи --------------------------------------- */

static const uint64_t KS_ZERO[16] = {
    UINT64_C(0x9864a848b6d4), UINT64_C(0x66845d0326d1),
    UINT64_C(0x9ce6eff8cd60), UINT64_C(0xde534c0aeec3),
    UINT64_C(0x84ed264ac5bb), UINT64_C(0xaedd6a826c93),
    UINT64_C(0x2c4d8a772d04), UINT64_C(0xaba7eec66f0f),
    UINT64_C(0xff822a71a117), UINT64_C(0x13776bbc1fbf),
    UINT64_C(0x499da61debdf), UINT64_C(0xff1503bd8aac),
    UINT64_C(0xdc17510f5fae), UINT64_C(0x3a08b3dab4f4),
    UINT64_C(0x63bbb6a9252b), UINT64_C(0x30a6e63f94e6)
};

static const uint64_t KS_SEQ[16] = {
    UINT64_C(0x352b543e645d), UINT64_C(0x3bb3b319fa45),
    UINT64_C(0xba34ee3b5816), UINT64_C(0xf5015d474faa),
    UINT64_C(0x18454389852b), UINT64_C(0x545daa0a6112),
    UINT64_C(0x51f0a63d8e60), UINT64_C(0xb9bddda2e38c),
    UINT64_C(0x4a92ed2d2816), UINT64_C(0x47cdd4e8e543),
    UINT64_C(0xe10755f1fc74), UINT64_C(0xb7f0a3bd6e0c),
    UINT64_C(0x3d5d558b3b7f), UINT64_C(0xe04b6a7bd904),
    UINT64_C(0x558c5a714b1b), UINT64_C(0xf9958a46e7bf)
};

static void test_key_schedule(void)
{
    uint8_t key[32];
    uint64_t rk[16];
    int i;

    printf("[8.2] subkeys for K=0\n");
    memset(key, 0, 32);
    dedalyan_key_schedule(key, rk);
    for (i = 0; i < 16; i++) ok_u64(rk[i], KS_ZERO[i], "k_i (K=0)");

    printf("[8.3] subkeys for K=000102..1e1f\n");
    for (i = 0; i < 32; i++) key[i] = (uint8_t)i;
    dedalyan_key_schedule(key, rk);
    for (i = 0; i < 16; i++) ok_u64(rk[i], KS_SEQ[i], "k_i (K=0001..1f)");
}

/* ---- раздел 8.4: векторы шифрования ------------------------------------ */

struct tv { const char *name, *p, *k, *c; };

static const struct tv TVS[4] = {
    { "TV1", "000000000000000000000000",
      "0000000000000000000000000000000000000000000000000000000000000000",
      "70a4ceaa4a6737fb294a0edf" },
    { "TV2", "000000000000000000000001",
      "0000000000000000000000000000000000000000000000000000000000000000",
      "85a903c2a6b73b50f00e1405" },
    { "TV3", "0123456789abcdef01234567",
      "000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
      "9b631fe623f15016cdba801e" },
    { "TV4", "ffffffffffffffffffffffff",
      "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
      "a936a309096319f869f6549d" }
};

static void test_vectors(void)
{
    int t, i;
    printf("[8.4] encryption vectors\n");
    for (t = 0; t < 4; t++) {
        uint8_t p[12], k[32], c[12], want[12], back[12];
        dedalyan_ctx ctx;
        hex2bytes(TVS[t].p, p, 12);
        hex2bytes(TVS[t].k, k, 32);
        hex2bytes(TVS[t].c, want, 12);
        dedalyan_key_setup(&ctx, k);
        dedalyan_encrypt_bytes(&ctx, p, c);
        checks++;
        if (memcmp(c, want, 12) != 0) {
            failures++;
            printf("  FAIL  %s ciphertext: got ", TVS[t].name);
            for (i = 0; i < 12; i++) printf("%02x", c[i]);
            printf(" want %s\n", TVS[t].c);
        }
        dedalyan_decrypt_bytes(&ctx, c, back);
        ok(memcmp(back, p, 12) == 0, "decrypt(encrypt(P)) == P");
    }
}

/* ---- раздел 8.5: пораундовые значения TV3 ------------------------------ */

static const uint64_t TV3_F[16] = {
    UINT64_C(0x229e7f0eb588), UINT64_C(0x94e01bc2aad8),
    UINT64_C(0xa9c66a1086e7), UINT64_C(0xfd9e1674955a),
    UINT64_C(0xa7c66b61b1b1), UINT64_C(0x82336d2325be),
    UINT64_C(0x05734be79f0d), UINT64_C(0x7e3febc6f8b6),
    UINT64_C(0x267030894057), UINT64_C(0xc572310f4b4c),
    UINT64_C(0x5f51ec67dd59), UINT64_C(0xbd7201f955d8),
    UINT64_C(0x60e553c6f23f), UINT64_C(0x1fb2103a188a),
    UINT64_C(0xaa69e031d8b8), UINT64_C(0x6f39670621ed)
};
static const uint64_t TV3_L[16] = {
    UINT64_C(0xcdef01234567), UINT64_C(0x23bd3a693c23),
    UINT64_C(0x590f1ae1efbf), UINT64_C(0x8a7b5079bac4),
    UINT64_C(0xa4910c957ae5), UINT64_C(0x2dbd3b180b75),
    UINT64_C(0x26a261b65f5b), UINT64_C(0x28ce70ff9478),
    UINT64_C(0x589d8a70a7ed), UINT64_C(0x0ebe4076d42f),
    UINT64_C(0x9defbb7feca1), UINT64_C(0x51efac110976),
    UINT64_C(0x209dba86b979), UINT64_C(0x310affd7fb49),
    UINT64_C(0x3f2faabca1f3), UINT64_C(0x9b631fe623f1)
};
static const uint64_t TV3_R[16] = {
    UINT64_C(0x23bd3a693c23), UINT64_C(0x590f1ae1efbf),
    UINT64_C(0x8a7b5079bac4), UINT64_C(0xa4910c957ae5),
    UINT64_C(0x2dbd3b180b75), UINT64_C(0x26a261b65f5b),
    UINT64_C(0x28ce70ff9478), UINT64_C(0x589d8a70a7ed),
    UINT64_C(0x0ebe4076d42f), UINT64_C(0x9defbb7feca1),
    UINT64_C(0x51efac110976), UINT64_C(0x209dba86b979),
    UINT64_C(0x310affd7fb49), UINT64_C(0x3f2faabca1f3),
    UINT64_C(0x9b631fe623f1), UINT64_C(0x5016cdba801e)
};

static void test_trace(void)
{
    uint8_t k[32];
    dedalyan_ctx ctx;
    dedalyan_block b;
    uint64_t tf[16], tl[16], tr[16];
    int i;

    printf("[8.5] per-round trace for TV3\n");
    hex2bytes(TVS[2].k, k, 32);
    dedalyan_key_setup(&ctx, k);
    b.l = UINT64_C(0x0123456789ab);
    b.r = UINT64_C(0xcdef01234567);
    dedalyan_encrypt_trace(&ctx, b, 16, tf, tl, tr);
    for (i = 0; i < 16; i++) {
        ok_u64(tf[i], TV3_F[i], "F_i");
        ok_u64(tl[i], TV3_L[i], "L_i");
        ok_u64(tr[i], TV3_R[i], "R_i");
    }
}

/* ---- раздел 10.4: точки, где вероятны ошибки --------------------------- */

static void test_pitfalls(void)
{
    dedalyan_ctx ctx;
    uint8_t key[32];
    uint64_t x;
    int i;

    printf("[10.4] pitfalls\n");

    /* 3. Сдвиг именно циклический, при s >= 48 берётся s mod 48. */
    x = UINT64_C(0x800000000001);
    ok_u64(dedalyan_rotl(x, 1), UINT64_C(0x000000000003), "rotl wraps (cyclic)");
    ok_u64(dedalyan_rotl(x, 48), x, "rotl by 48 == identity");
    ok_u64(dedalyan_rotl(x, 49), dedalyan_rotl(x, 1), "rotl s mod 48");
    ok_u64(dedalyan_rotr(dedalyan_rotl(x, 13), 13), x, "rotr undoes rotl");
    ok_u64(dedalyan_rotl(x, 0), x, "rotl by 0 == identity");

    /* 2. Маскирование после умножения: результат всегда 48-битный. */
    for (i = 0; i < 64; i++) {
        uint64_t v = dedalyan_f(UINT64_C(0xFFFFFFFFFFFF) - (uint64_t)i,
                                UINT64_C(0xFFFFFFFFFFFF), (unsigned)i & 15u);
        ok(v <= DEDALYAN_MASK, "F output fits in 48 bits");
    }

    /* 7. Расшифровка на урезанном числе раундов тоже обратима. */
    memset(key, 0x5a, 32);
    dedalyan_key_setup(&ctx, key);
    for (i = 1; i <= 16; i++) {
        dedalyan_block p, c, d;
        p.l = UINT64_C(0x0123456789ab);
        p.r = UINT64_C(0xcdef01234567);
        c = dedalyan_encrypt_r(&ctx, p, (unsigned)i);
        d = dedalyan_decrypt_r(&ctx, c, (unsigned)i);
        ok(d.l == p.l && d.r == p.r, "round-reduced roundtrip");
    }
}

/* ---- обратимость и CTR -------------------------------------------------- */

static void test_roundtrip_random(void)
{
    uint64_t st = UINT64_C(0xD1CE5EED);
    dedalyan_ctx ctx;
    size_t i;

    printf("[extra] random roundtrip, 200000 pairs\n");
    for (i = 0; i < 200000; i++) {
        uint8_t key[32];
        dedalyan_block p, c, d;
        int j;
        for (j = 0; j < 4; j++) {
            uint64_t v = ded_k_splitmix64(&st);
            int b;
            for (b = 0; b < 8; b++) key[j * 8 + b] = (uint8_t)(v >> (56 - 8 * b));
        }
        p.l = ded_k_splitmix64(&st) & DEDALYAN_MASK;
        p.r = ded_k_splitmix64(&st) & DEDALYAN_MASK;
        dedalyan_key_setup(&ctx, key);
        c = dedalyan_encrypt(&ctx, p);
        d = dedalyan_decrypt(&ctx, c);
        if (d.l != p.l || d.r != p.r) {
            failures++;
            printf("  FAIL  roundtrip at sample %llu\n", (unsigned long long)i);
            break;
        }
    }
    checks++;
}

static void test_ctr(void)
{
    dedalyan_ctx ctx;
    uint8_t key[32], counter[12], pt[100], ct[100], back[100];
    int i;

    printf("[extra] CTR mode\n");
    for (i = 0; i < 32; i++) key[i] = (uint8_t)(i * 7 + 1);
    memset(counter, 0, 12);
    counter[11] = 0xFD;          /* проверяем перенос через границу байта */
    for (i = 0; i < 100; i++) pt[i] = (uint8_t)i;

    dedalyan_key_setup(&ctx, key);
    dedalyan_ctr(&ctx, counter, pt, ct, sizeof(pt), 16);
    dedalyan_ctr(&ctx, counter, ct, back, sizeof(ct), 16);
    ok(memcmp(pt, back, sizeof(pt)) == 0, "CTR is self-inverse");
    ok(memcmp(pt, ct, sizeof(pt)) != 0, "CTR actually changes data");

    /* Счётчик = все единицы: инкремент должен завернуться в ноль. */
    memset(counter, 0xFF, 12);
    dedalyan_ctr(&ctx, counter, pt, ct, sizeof(pt), 16);
    dedalyan_ctr(&ctx, counter, ct, back, sizeof(ct), 16);
    ok(memcmp(pt, back, sizeof(pt)) == 0, "CTR wraps 2^96 correctly");
}

int main(void)
{
    printf("Dedalyan test vectors -- %s\n\n", dedalyan_version());

    test_round_constants();
    test_labyrinth();
    test_key_schedule();
    test_vectors();
    test_trace();
    test_pitfalls();
    test_roundtrip_random();
    test_ctr();

    printf("\n%d checks, %d failures\n", checks, failures);
    if (failures == 0) printf("ALL TESTS PASSED\n");
    return failures ? 1 : 0;
}
